### PR TITLE
fix(challenge): account for browser zoom, add note to challenge (for width challenge)

### PR DESCRIPTION
**Pull request by lasjorg**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

This is the same fix as for the [height challenge](https://github.com/freeCodeCamp/freeCodeCamp/pull/37687). I do not have any confirmation for this fix yet ([forum post](https://www.freecodecamp.org/forum/t/is-the-adjust-the-width-lesson-in-the-applied-vis-design-course-broken/324794) showing the issue). We may or may not want to wait for one.

I thought about wrapping this in a function to add a variable explaining the `- 10` part, just to avoid any "magic" numbers. But that just makes the assert really long and without proper formatting it gets hard to read in my opinion.

The `- 10` is subtracting the padding and border of the `.fullCard` element to get the "true" width.

### CODE REVIEW

**Comment by ojeytonwilliams** [comment]
@lasjorg It's possible to format the YAML strings however you want, but you need to wrap them in single or double quotes.  Here's the [spec](https://yaml.org/spec/1.2/spec.html), but the main points are 

- `"` you have to escape `"` and `\` with `\`
- `'` you have to repeat  `'`

For example:
```yaml
tests:
  - text: Your code should change the <code>width</code> property of the card to 245 pixels by using the <code>fullCard</code> class selector.
    testString: "assert(
  Math.round(
    document.querySelector('.fullCard').getBoundingClientRect().width
  ) -
    10 ===
    245 &&
    /\\.fullCard{\\S*width:245px(;\\S*}|})/.test(
      $('style')
        .text()
        .replace(/\\s/g, '')
    )
);"
```

**Comment by RandellDawson** [comment]
@lasjorg You could also create a helper function and place it in the `After Test` section and call the helper function in the assert.  Then add any applicable comments inside the helper function.

**Comment by lasjorg** [comment]
Thank you both for the info.

I don't think I have ever seen a challenge where the assert code was formatted. I like the formatting but I think the escaping makes the regex harder to read and is likely to confuse some if unaware of the YAML rules.


@ojeytonwilliams @RandellDawson Do you guys think I need to make it more clear what the `- 10` is, or should i just leave it as is?

**Comment by lasjorg** [comment]
Switched to just using regex.

File: curriculum/challenges/english/01-responsive-web-design/applied-visual-design/adjust-the-width-of-an-element-using-the-width-property.english.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ img {
 ## Instructions
 <section id='instructions'>
 Add a <code>width</code> property to the entire card and set it to an absolute value of 245px. Use the <code>fullCard</code> class to select the element.
+<strong>Note:</strong> You may need to be at 100% zoom to pass the test on this challenge.
```
Comment (ojeytonwilliams) [inline curriculum/challenges/english/01-responsive-web-design/applied-visual-design/adjust-the-width-of-an-element-using-the-width-property.english.md]:
```suggestion
```
No longer needed.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Looks good.  Just need to remove the warning.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Comment by RandellDawson** [comment]
Are there any outstanding issues that this PR closes?  If so, can we get them listed in the PR body?

**Comment by lasjorg** [comment]
@RandellDawson I don't think there are any open issues for this challenge (just the [one closed](https://github.com/freeCodeCamp/freeCodeCamp/issues/37695)). At least not that I found.

**Review by RandellDawson — CHANGES_REQUESTED** [review]
I can pass the challenge with the following which do not actually affect the width.

```css
  .fullCard {
    border: 1px solid #ccc;
    border-radius: 5px;
    margin: 10px 5px;
    padding: 4px;
    max-width: 245px;
  }
```

OR
```css
  .fullCard {
    border: 1px solid #ccc;
    border-radius: 5px;
    margin: 10px 5px;
    padding: 4px;
    Width: 245px;
  }
```
OR
```css
  .fullCard {
    border: 1px solid #ccc;
    border-radius: 5px;
    margin: 10px 5px;
    padding: 4px;
    mywidth: 245px;
  }
```

**Comment by lasjorg** [comment]
@RandellDawson Yes I know, sorry. I did see that the regex(s) was lax but I just kept it like it was. It also allows for incorrect syntax (also because of the `\S*`).

```
.fullCard {{{{}}}
  border: 1px solid #ccc;
  border-radius: 5px;
  margin: 10px 5px;
  padding: 4px;;;
  width: 245px;;;;
}}}}
```

I guess we have to rewrite them. If anyone has any good suggestions I'm all ears.

Not going to lie. I'm half tempted to just put back the zoom note and leave it how it was to begin with. This zoom nonsense is apparently a ["feature" of Blink](https://bugs.chromium.org/p/chromium/issues/detail?id=359691) and not a bug.

Who knows just how many challenges are affected. Rewriting 3 or 4 challenges is not so bad, but if we have to rewrite a bunch of them just to accommodate this one problem then I'm not so sure it's worth it. Adding a note to each affected challenge might make more sense. That way we also won't inadvertently introduce new bugs.

I'm just confused why we haven't seen this reported before now? These (new?) bug reports seems to oddly correspond in time with the latest major site update. But maybe it's just a coincidence.

**Comment by ojeytonwilliams** [comment]
This is why I thought we'd better off without regex.  It accepts both correct ways of doing it (I checked that much), but so much more besides (which I should have checked, but didn't).  Apparently we have to.

As for solving this, aren't there two types of tests that have the bug. Dimensions set inline and dimensions set by css?  The first type seems to be solved by checking `.style` (like you (@lasjorg) did in https://github.com/freeCodeCamp/freeCodeCamp/pull/37756).  The second type is the real problem, but how about we try to come up with a robust regex for extracting a css rule and apply that to all the relevant challenges?

> Adding a note to each affected challenge might make more sense.

Except that that is a pain for people that actually need it to be zoomed to be able to read.

> This zoom nonsense is apparently a "feature" of Blink and not a bug.

Eh, I don't agree with them.  Since they aren't consistent with Firefox, at least one of them has a bug.

**Comment by lasjorg** [comment]
I have reverted to my initial code for now.

@ojeytonwilliams Sure some fixes are pretty easy. I'm just not sure how many challenges are actually affected by this.

> Eh, I don't agree with them. Since they aren't consistent with Firefox, at least one of them has a bug.

It seems they have decided that precision of 1/64th of a pixel is enough. I guess it might be a performance thing.


Unless we can come up with a good regex, I think my initial code for the PR is the best we have so far. I don't see the rounding being a problem here because we are limiting the valid input to 245 using the regex so the user can not enter some weird value. Also, this is on a fixed sized element so the viewport dimension does not dynamically change the element dimension. These two facts combined makes me think we can get away with rounding the number.

Theoretical issues aside. I think we should implement it and get it tested, at least until we have something better. It can't be worse than what we have in production now.

**Comment by moT01** [comment]
I've used a regex like this a number of times in the past @lasjorg... 
```
const fullCard = code.match(/\.fullCard\s*{[\s\S]+?[^}]}/g)[0].toLowerCase(); 
assert(/width\s*:\s*245pxs*(;|})/g.test(fullCard));
```
not sure if you want to keep changing this around, but it's pretty robust - it could apply to the other PR as well.

**Comment by lasjorg** [comment]
@moT01 @ojeytonwilliams Do you think we need to use the better regex or should we just go with the one we already have?

**Review by moT01 — CHANGES_REQUESTED** [review]
The current state of this PR doesn't address @RandellDawson's concerns, they are still a problem - although his second example should pass since capitalization still applies the style. 

After playing around with this for a little bit, the rounding issue was just too much of a pain for me. I came up with this...
```
    testString: const fullCard = code.match(/\.fullCard\s*{[\s\S]+?[^}]}/g)[0];
assert(/width\s*:\s*245px\s*(;|})/gi.test(fullCard) && 
getComputedStyle(document.querySelector('.fullCard')).maxWidth === 'none');
```

It's similar to what I suggested earlier, but adds a check that `max-width` isn't used. That's what I would go with, but I don't really care as long as it works and, right now, it isn't.

**Comment by lasjorg** [comment]
@moT01 Thank you for the help.

There are edge cases for broken syntax with this pure regex based approach. You can break the output while still passing the test.

Invalid syntax after the opening brace:

```
.fullCard {{}
  width: 245px;
  border: 1px solid #ccc;
  border-radius: 5px;
  margin: 10px 5px;
  padding: 4px;
}
```

Extra character before `width`:

```
.fullCard {
  swidth: 245px;
  border: 1px solid #ccc;
  border-radius: 5px;
  margin: 10px 5px;
  padding: 4px;
}
```

Missing semicolon on the rule that comes before:

```
.fullCard {
  border: 1px solid #ccc;
  border-radius: 5px;
  margin: 10px 5px;
  padding: 4px
  width: 245px;
}
```

There are probably other edge cases I haven't considered. Not sure how, or if, we should test for this?

**Comment by ojeytonwilliams** [comment]
@lasjorg I think any regex we come up with is likely to accept some incorrect answers.  As long as it passes all the *correct* answers we can come up with and as long as it rejects typical mistakes, I'm okay with that.  As long as @RandellDawson and @moT01 are happy, that's good enough for me.

**Comment by lasjorg** [comment]
Just so I don't forget, I have push this "as is" for now.

1. Made changes to avoid throwing an error if `.match()` returns null.

2. Used jQuery for the `maxWidth` check just to shorten the assert a bit.

**Review by moT01 — APPROVED** [review]
Alright, I'm fine adding it like this, it fixes the rounding issues that were the main problem. The only edge case that I found is that using `min-width: 245px` will pass. But, that input is unlikely and I hate to keep asking for changes for small things like that. If any other edge cases come up, we can deal with them then. Thanks for all the work on this @lasjorg 🎉

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
Thank you for your contribution to the page! 👍
We're happy to accept these changes, and look forward to future contributions. 🎉
